### PR TITLE
Issue 80 refine table editing controller design

### PR DESCRIPTION
**Pull request by psartzetakis**
## Pull request checklist

- [x] All tests pass. 
- [x] Demo project builds and runs.
- [x] I have resolved merge conflicts.
- [x] I have followed the [coding style](https://github.com/jessesquires/HowToContribute#style-guidelines) and [contributing guidelines](https://github.com/jessesquires/HowToContribute).

## What's in this pull request?

1. The `Item` now is passed on the `CanEditRow` closure.
2. The `dataSource` is now passed on the `CommitEditingStyleForRow` closure.
3. The user can now pass the `tableEditingController` closure directly on the initialiser.
4. Added example for the `FetchedResultsController`

### CODE REVIEW

File: Source/DataSourceProvider.swift [hunk 1]
```diff
@@ -37,32 +37,34 @@ where CellFactory.Item == DataSource.Item, SupplementaryFactory.Item == DataSour
...
+     - parameter tableEditingController: The tableEditing controller.
+     
      - returns: A new `DataSourceProvider` instance.
 
      - warning: Table views do not have supplementary views, but this parameter is still required in order to satisfy
      the generic constraints for Swift. You can simply pass the same `cellFactory` here. The parameter will be ignored.
      The same applies to collection views that do not have supplementary views. Again, the parameter will be ignored.
      */
-    public init(dataSource: DataSource, cellFactory: CellFactory, supplementaryFactory: SupplementaryFactory) {
+    public init(dataSource: DataSource, cellFactory: CellFactory, supplementaryFactory: SupplementaryFactory, tableEditingController: TableEditingController<DataSource>? = nil) {
```
Comment (psartzetakis) [inline Source/DataSourceProvider.swift:58]: With the current update it is possible for the user to pass the `tableEditingController ` through the initialiser. However I have still left the old implementation with the `_tableEditingController`  for the time being. I was thinking of deprecating the old way (assigning it through the variable`tableEditingController`) but I wanted your opinion first @jessesquires .

**Comment by psartzetakis** [comment]
Hi @jessesquires ,
Apologies for the delay..😅. I finally managed to submit the PR 😀.

**Comment by jessesquires** [comment]
thanks @psartzetakis ! sorry for the delay 😄 

i'll try to get this and #98 merged soon.

-----

i've decided to drop the `6.1.0 milestone` in favor of just doing `7.0.0` with Swift 4 and other breaking-changes refactoring 😄

**Comment by jessesquires** [comment]
hey @psartzetakis ! finally merged 🎉 

this is great. 😎 

I made a couple of minor edits in 14d4df0dceecf1150e7015f0630fee448de8a15e:

- Like you suggested above, remove the public var for `tableEditingController`. You can only pass it via the `init`.
- I moved the `init` that receives a `tableEditingController ` to the extension that's constrained *only for table view use*. This way, you can't pass this in for a collection view data source. 😄 
https://github.com/jessesquires/JSQDataSourcesKit/blob/develop/Source/DataSourceProvider.swift#L65-L84

**Comment by jessesquires** [comment]
I'm hoping to get 7.0 release by Monday, Jan 1. 🎉 

also cc @dcaunt 😄

**Comment by psartzetakis** [comment]
Hey @jessesquires , Awesome news  🎉

**Pull request merged**